### PR TITLE
[RFC]Shake up RollMeleeOutcomeAgainst

### PR DESCRIPTION
**Pull request by Phatcat**
This change is [Reviewable](https://reviewable.io/reviews/cmangos/mangos-classic/161)

### CODE REVIEW

**Comment by Phatcat** [comment]
@VladimirMangos @Oniryck

One of you guys maybe up for reviewing this?

I'm not quite sure how to test it properly, any input on that would be appreciated.

**Comment by Phatcat** [comment]
Anyone at all?

**Comment by cala** [comment]
I mostly played casters, so I could not give you proper feedback on this, I'm afraid.

**Comment by dagochen** [comment]
I dont understand why you reroll for every case. Current master did the rolling exactly as people described the combat tables to be working and changing master to be different from that is maybe harder to recognize? Even though multiple rolls dont affect the chances here, it might have sideeffects?

**Comment by Phatcat** [comment]
You are indeed right that there are some non-blizzlike changes that need to be addressed; I made these changes using common sense when I should have realized that Blizzard has none.

I'm pretty sure the multiple rolls actually do affect the chances, which I've probably misunderstood since the wiki and other sources very well say that changing one chance affects the others.

This whole PR is stale. I'll either freshen it up or close it down.

**Comment by cyberium** [comment]
@Phatcat its mostly up to you, if you feel this PR will realy improve something in combat feeling then merge it. You have to go further following your idea.
If you are starting to doubt about it then there is maybe a reason. So either reevaluate it or abandon it.
I personaly think that this code is a bit more clear in your version.

**Comment by Phatcat** [comment]
Yes, well... I think that some of it would, but if it's towards a less blizzlike state.

Right now my thought is that my PR does improve on some things, but it may very well be at the cost of unbalancing percentages or introducing unblizzlike behaviour in others.

this PR very much needs and overhaul / to get shut down before moving further with this.
sorry for asking people for input on this when really what it needs most is my own input currently.

**Comment by dagochen** [comment]
I haven't checked every case, but I think it does not change chances at all or? Only theoretical, because we reroll. But every roll has the same quality of randomness, it should not change that much or?

**Comment by Phatcat** [comment]
that randomness, statistical part area of math is something I never bothered much with sadly.. I'll look into this one soonish, have a few things I have to take care of first.

**Comment by Phatcat** [comment]
the thing is it shouldn't be as random as one would think, so I changed it somewhat back to the old way.

I would love for someone to give some feedback on this, as I finally feed this is somewhat ready.

File: src/game/Unit.cpp [hunk 1]
```diff
@@ -2201,132 +2201,110 @@ MeleeHitOutcome Unit::RollMeleeOutcomeAgainst(const Unit* pVictim, WeaponAttackT
...
-    }
 
-    // block chances
-    // check if attack comes from behind, nobody can parry or block if attacker is behind
-    if (!from_behind)
-    {
-        if (pVictim->GetTypeId() == TYPEID_PLAYER || !(((Creature*)pVictim)->GetCreatureInfo()->ExtraFlags & CREATURE_EXTRA_FLAG_NO_BLOCK))
+        tmp = block_chance - skillBonus;
+        // check if unit _can_ block
+        if (roll < (sum += tmp) && (pVictim->GetTypeId() == TYPEID_PLAYER
```
Comment (dagochen) [inline src/game/Unit.cpp]: I would put these conditions up to the start where we decide if someone can parry (similar to the dodge part).

File: src/game/Unit.cpp [hunk 2]
```diff
@@ -2201,132 +2201,110 @@ MeleeHitOutcome Unit::RollMeleeOutcomeAgainst(const Unit* pVictim, WeaponAttackT
...
-        int32 maxskill = attackerMaxSkillValueForLevel;
-        skill = (skill > maxskill) ? maxskill : skill;
-
-        tmp = (10 + 2 * (victimDefenseSkill - skill)) * 100;
-        tmp = tmp > 4000 ? 4000 : tmp;
-        if (roll < (sum += tmp))
+        tmp = parry_chance - skillBonus;
+        // check if unit _can_ parry
+        if (roll < (sum += tmp) && (pVictim->GetTypeId() == TYPEID_PLAYER
+            || !(((Creature*)pVictim)->GetCreatureInfo()->ExtraFlags & CREATURE_EXTRA_FLAG_NO_PARRY)))
```
Comment (dagochen) [inline src/game/Unit.cpp]: I would put these conditions up to the start where we decide if someone can parry (similar to the dodge part).
 Add a line note

File: src/game/Unit.cpp [hunk 3]
```diff
@@ -2201,132 +2201,110 @@ MeleeHitOutcome Unit::RollMeleeOutcomeAgainst(const Unit* pVictim, WeaponAttackT
...
     {
+        roll = urand(0, 100) * 100;
         uint32 typeId = GetTypeId();
-        if ((typeId == TYPEID_UNIT && !(GetOwnerGuid() && GetOwner()->GetTypeId() == TYPEID_PLAYER)
+        tmp = (victimDefenseSkill < victimMaxSkillValueForLevel) ? victimDefenseSkill : victimMaxSkillValueForLevel;
+
+        if (roll < (tmp = (((attackerMaxSkillValueForLevel - tmp) * 200) - 1500))
+            && ((typeId == TYPEID_UNIT && !(GetOwnerGuid() && GetOwner()->GetTypeId() == TYPEID_PLAYER)
             && !(((Creature*)this)->GetCreatureInfo()->ExtraFlags & CREATURE_EXTRA_FLAG_NO_CRUSH))
-            || (typeId == TYPEID_PLAYER && GetCharmerGuid() && GetCharmer()->GetTypeId() == TYPEID_UNIT))
```
Comment (dagochen) [inline src/game/Unit.cpp]: I would put these conditions up to the start where we decide if someone can parry (similar to the dodge part).
 Add a line note

**Comment by dagochen** [comment]
I did some comments for consistency. You can seperate the if-statements because they are 2 different things. One thing belongs to our use case (like "can a player or creature dodge, parry, block....), while the roll itself is the technical way to achieve it (what was chance and the roll).

**Comment by Phatcat** [comment]
> I did some comments for consistency.

Thank you.

> You can seperate the if-statements because they are 2 different things.

I'm not entirely sure I understand what you mean by this, but if you mean that I can nest the roll check inside the if statement for the 'conditioned' checks. Well, I tend to not like that style at all, also it is slightly less optimized than keeping/simplifying it into just one if statement. I will however agree with you about your changes suggested in your comments, the question is however if the int comparison or the conditioned checks are most lightweight on the server, because we should always put the heaviest work last to avoid it as much as possible, obviously.

File: src/game/Unit.cpp [hunk 4]
```diff
@@ -2201,132 +2201,110 @@ MeleeHitOutcome Unit::RollMeleeOutcomeAgainst(const Unit* pVictim, WeaponAttackT
...
-        && roll < (tmp = (((attackerMaxSkillValueForLevel - tmp) * 200) - 1500)))
+    else if ((getLevel() - 3) >= pVictim->getLevel() && !SpellCasted)
     {
+        roll = urand(0, 100) * 100;
         uint32 typeId = GetTypeId();
-        if ((typeId == TYPEID_UNIT && !(GetOwnerGuid() && GetOwner()->GetTypeId() == TYPEID_PLAYER)
+        tmp = (victimDefenseSkill < victimMaxSkillValueForLevel) ? victimDefenseSkill : victimMaxSkillValueForLevel;
+
+        if (roll < (tmp = (((attackerMaxSkillValueForLevel - tmp) * 200) - 1500))
+            && ((typeId == TYPEID_UNIT && !(GetOwnerGuid() && GetOwner()->GetTypeId() == TYPEID_PLAYER)
```
Comment (Phatcat) [inline src/game/Unit.cpp]: probably move much of these 'conditioned' checks up to where they belong, with the other conditions up in the 'else if' at line 2298, and actually only keep the roll check here

**Comment by dagochen** [comment]
If you just focus on performance you might wanna change some of the rolls to the beginning of the if-statements, because parry and block have usually lower chance and will make the entire condition be false, which would make the more expensive typecasts like 
`Line 242: !(((Creature*)pVictim)->GetCreatureInfo()->ExtraFlags & CREATURE_EXTRA_FLAG_NO_PARRY))`
obsolete.

I am not sure if you understood what I was talking about. It was about SoC. At this point you can seperate the logic from how we achieve it, the technical part. 
Yours: https://gist.github.com/dagochen/d0000eb18e2a1c110b3ef75241d7385f
Mine: https://gist.github.com/dagochen/3f8d25a4ec3ac9879ab167e445736022

**Comment by Phatcat** [comment]
> If you just focus on performance you might wanna change some of the rolls to the beginning of the if-statements

Exactly like I had done, which you told me to change...

> You can seperate the if-statements because they are 2 different things.
(...)
I am not sure if you understood what I was talking about.

I understood perfectly fine what you were talking about, I even replied to it. What you're talking about, I already told you I did not like seperating the if statements like that. also if you want to display what I'm doing do it properly, because this, is mine: 
https://gist.github.com/Phatcat/27981bfba767de2c2a8959d6bcbcbd95
My !from_behind covers not only parry, but also block. How do you then want me to do put block/parry specific conditions in a check that is meant to check the possibility for BOTH block and parry?

I think you may want to re-read my functions because I think you got the logic a bit wrong on your end.

**Comment by dagochen** [comment]
You missunderstand it again...like always...I should really give up on you....

**Comment by Phatcat** [comment]
. . . 
> Yours: https://gist.github.com/dagochen/d0000eb18e2a1c110b3ef75241d7385f

That's not how the function looks in my work. It looks like this
https://gist.github.com/Phatcat/27981bfba767de2c2a8959d6bcbcbd95

the scope of !from_behind does NOT stop after parry, it carries on all the way down to after block....

Here is where you seem to think it stops https://gist.github.com/Phatcat/27981bfba767de2c2a8959d6bcbcbd95#file-gistfile1-txt-L11

Here is where it actually stops https://gist.github.com/Phatcat/27981bfba767de2c2a8959d6bcbcbd95#file-gistfile1-txt-L29

That's not the same. Nor does it allow for setting it up like you're showing, that would require me to do !from_behind individually on both parry and block, which I havn't, I do one collectively for both because that's more simple... You could have suggested that I change that (I would have said no, though). Is that what you're suggesting? Because no then.

Also you pack away that attitude and whatever grudge you got going on against me. There's lots of other stuff to do around cmangos instead of trying to pick fights with me. I am sick of the drama. Thank you.

**Comment by dagochen** [comment]
You simply never understand what people want from you, can't help on that. Maybe my example was misleading, sorry for that.
Pros on changing it the way I told you:
+ Readability
+ Seperation of Concerns
+ Performance on if-statements increase, because you do the more probable booleans first

Cons:
- 4 instead of 3 if-statements (2 for parry and block, with 2 nested inside)

**Comment by Phatcat** [comment]
One more con, doing the same logical operation twice (checking from_behind)

Now, performance on the other hand, let's talk about that one. By keeping less if statements I increase performance; inside these if statements we can place the conditions however we choose. If you say that the roll check should be placed first due to it being the lightest (I won't argue here, I don't know, I have been asking for clearification on which operations were the heaviest for quite some time), my argument however goes back to; that was what I had initially done, where you told me to change it.

That's... Nice.

**Comment by Phatcat** [comment]
> You simply never understand what people want from you, can't help on that. Maybe my example was misleading, sorry for that.

That comment is hilariously double-sided by the way ^^ - maybe stop misleading people if you don't want them to misunderstand you another time.

Merging this one soon, getting tired of looking at this PR and even though it's nowhere near as aggressive as at first, it still should solve a few minor things.

**Pull request closed without merging**